rfcs/20230110-kernel-reuse/README.md

```diff
@@ -0,0 +1,265 @@
+# Proposal for Primitive Compilation in oneDNN
+
+## 1. Introduction
+Slow primitive compilation consistently causes customer issues in oneDNN. To
+handle this, we introduced a [global primitive
+cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200207-global-primitive-cache).
+This solution is insufficient for the GPU software stack, especially for OpenCL
```

Comment:
Can you please elaborate on this a little bit? Basically, there are 3 use cases that I'm aware of:
1. OpenVINO use case. Framework creates primitives only once and then re-uses them. They reduces creation overhead by using persistent cache API. They don't rely on oneDNN primitive cache
2. PyTorch and TensorFlow use case. They create primitives for each call of an operation. The primitive cache is filled during the 1st iteration and then the primitives are re-used
3. In some cases frameworks use different problem sizes for each iteration which makes the primitive cache inefficient.

As far as I understand we reduced creation overhead for the first 2 uses cases. However, the 3rd one is still a problem. Given that, I think the problem statement should be focused on significant overhead for dynamic shapes specifically.

Reply:
Ok, I can expand on this, a couple of quick points.
>OpenVINO use case. Framework creates primitives only once and then re-uses them. They reduces creation overhead by using persistent cache API. They don't rely on oneDNN primitive cache.

To some extent, this is OpenVino making something work despite our libraries behavior. They also implement parallel kernel creation to workaround this too Our goal should be to have reasonable out of box behavior, not just provide tools to workaround poor behavior. As such, I am would not really categorize this as solving the problem. We are just providing a workaround.

> PyTorch and TensorFlow use case. They create primitives for each call of an operation. The primitive cache is filled during the 1st iteration and then the primitives are re-used

While this massively helps the issue, it is not enough. For example, we are by default compiling custom OpenCL kernels for each eltwise primitive. Some of these operations take around 0.04 ms to execute on the GPU. The compilation process takes about 500 ms. Because of this, our customers need to reuse that specific eltwise primtive 300 / 0.04 = 12,500 for compute time to even match the time invested on the CPU generating the primitive. Even if it hasn't been reported, it is *extremely* unlikely this is not a problem for some scenario. Preferably, we would not even call the OpenCL compiler for such short lived operations, but we need to resolve kernel reuse issues before we worry about alternative kernel distribution methods. While I don't think this is the most important issue, we should keep this first compilation issue in mind while designing a solution.

> In some cases frameworks use different problem sizes for each iteration which makes the primitive cache inefficient.

You are correct, this is the most pressing use case.

Reply:
> To some extent, this is OpenVino making something work despite our libraries behavior. They also implement parallel kernel creation to workaround this too Our goal should be to have reasonable out of box behavior, not just provide tools to workaround poor behavior. As such, I would not really categorize this as solving the problem. We are just providing a workaround.

I don't think I agree, the overhead is coming from using OpenCL programming model. There is nothing we can do to reduce compilation time if the kernels were not compiled before and stored in some cache so I don't think this is just a workaround. Another important thing to mention is that frameworks may define reasonable out-of-the-box behavior differently. For example, PyTorch and TF are fine with the primitive cache that is enabled out-of-the-box though OpenVINO doesn't need it but they need another feature which is persistent cache.

What I'm trying to say is that the problem seems to be solely about dynamic shapes to me and it should be investigated as such. To me this is a discussion about enabling runtime dimensions vs enabling kernel cache.